File: 1_algorithmic_toolbox/week4/cpp/2_binary_search_duplicates.cpp
#include <iostream>
#include <vector>

int binary_search_duplicates(const std::vector<int> &sequence, const int query)
{
  int left { 0 }, right { static_cast<int >(sequence.size()) }, mid {};
  while (mid = left + (right - left) / 2, left < right)
    if (sequence[mid] == query && (mid == 0 || sequence[mid - 1] != query))
      return mid;
    else if (sequence[mid] >= query)
      right = mid;
    else
      left = mid + 1;

  return -1;
}

int main()
{
  int n {}, m {}, query {};
  std::cin >> n;
  std::vector<int> sequence(n);
  for (auto &e : sequence) std::cin >> e;
  std::cin >> m;
  for (int i { 0 }; i < m; ++i) {
    std::cin >> query;
    std::cout << binary_search_duplicates(sequence, query) << ' ';
  }
  std::cout << std::endl;
}

File: 1_algorithmic_toolbox/week4/python/2_binary_search_duplicates.py
def binary_search_duplicates(seq, query):
    (left, right) = (0, len(seq))
    while left < right:
        mid = left + (right - left) // 2
        if seq[mid] == query and (mid == 0 or seq[mid - 1] != query):
            return mid
        elif seq[mid] >= query:
            right = mid
        else:
            left = mid + 1

    return -1


if __name__ == "__main__":
    input()  # skip length of sequence
    seq = [int(e) for e in input().split()]
    input()  # skip number of queries
    for q in input().split():
        print(binary_search_duplicates(seq, int(q)), end=" ")
    print()
